File: elixir/lib/day04.ex
defmodule Aoc.Day04 do
  @moduledoc """
  Day 4: Camp Cleanup.
  """

  @input_file "../data/input04.txt"
  @sample_file "../data/input04-sample.txt"

  def read_input_data, do: File.read!(@input_file)
  def read_sample_data, do: File.read!(@sample_file)
  def split_lines(data), do: String.split(data, "\n", trim: true)

  # Main

  def main do
    IO.inspect({
      part_one(),
      part_two()
    })
  end

  # Part One

  def part_one do
    read_input_data()
    |> split_lines()
    |> Enum.map(&split_map_set/1)
    |> Enum.map(&subsets?/1)
    |> Enum.filter(& &1)
    |> Enum.count()
  end

  def subsets?([left, right]) do
    MapSet.subset?(left, right) or
      MapSet.subset?(right, left)
  end

  # Part Two

  def part_two do
    read_input_data()
    |> split_lines()
    |> Enum.map(&split_map_set/1)
    |> Enum.map(&intersect?/1)
    |> Enum.filter(& &1)
    |> Enum.count()
  end

  def intersect?([left, right]) do
    not MapSet.disjoint?(left, right)
  end

  # Private

  defp split_map_set(line) do
    line
    |> String.split(",", trim: true)
    |> Enum.map(fn range ->
      [first, last] =
        range
        |> String.split("-", trim: true)
        |> Enum.map(&String.to_integer/1)

      first
      |> Range.new(last)
      |> MapSet.new()
    end)
  end
end

File: haskell/Day04.hs
module Day04 where
import Data.List.Split (splitOn)
import System.IO.Unsafe

inputData :: String
inputData = unsafePerformIO . readFile $ "../data/input04.txt"

sampleData :: String
sampleData = unsafePerformIO . readFile $ "../data/input04-sample.txt"

type Range = (Int, Int)

anyOverlap :: Range -> Range -> Bool
anyOverlap (a, b) (x, y)
  | x < a = anyOverlap' (x, y) (a, b)
  | otherwise = anyOverlap' (a, b) (x, y)

anyOverlap' :: Range -> Range -> Bool
anyOverlap' (a, b) (x, y)
  | a <= x && b >= x = True
  | b >= y = True
  | otherwise = False

eitherContained :: Range -> Range -> Bool
eitherContained (a, b) (x, y)
  | a >= x && b <= y  = True
  | a <= x &&  b >= y = True
  | otherwise = False

parseRanges :: String -> [Range]
parseRanges line = ranges
  where dashedRanges = splitOn "," line              -- ["2-4","6-8"]
        rangePairs = map (splitOn "-") dashedRanges  -- [["2","4"],["6","8"]]
        ranges = map parseRange rangePairs           -- [(2,4),(6,8)]

parseRange :: [String] -> Range
parseRange [left, right] = (read left, read right)
parseRange _ = error "range requires two elements"

parsePair :: (Range -> Range -> Bool) -> [Range] -> Bool
parsePair f [leftPair, rightPair] = f leftPair rightPair
parsePair _ _ = error "pair requires two ranges"

-- Main

main :: IO ()
main = print (total1, total2)
  where inputLines = lines inputData
        ranges = map parseRanges inputLines
        contains = map (parsePair eitherContained) ranges
        -- overlaps = map (\r -> (r, parsePair anyOverlap r)) ranges
        overlaps = map (parsePair anyOverlap) ranges
        total1 = length (filter id contains)
        total2 = length (filter id overlaps)
